### ISSUE TITLE
Creating logic-only components

### ISSUE THREAD

**Issue by mrozbarry** [facebook/react#4885]
I'm developing an app and want to dynamically load content into my application.  I am using flux, so there is a fairly defined pipeline for getting and managing data.

What I want is to do something like this:

```jsx
<MyRootComponent ...>
  <LoadSomethingInteresting ...>
    <ComponentThatUsesLoadedData />
  </LoadSomethingInteresting>
</MyRootComponent>
```

Currently, I get around this by using `<div>` as the rendered component, but this doesn't work well when I lazy load into a table:

```jsx
<MyRoot>
  <table>
    <thead>
      <LoadSomeTheadData>
        <ComponentThatRendersTrAndTds />
      </LoadSomeTheadData>
    </thead>
  </table>
</MyRoot>
```

Chome (but probably in other browsers) will push the loading div outside the table, which will force the browser to inject it's own corrected HTML, and React will lose it's root since the DOM gets modified without React's consent.

So is there any way I can do something similar to:

```javascript
  render: function() {
    return null; // won't work because this.props.children won't get rendered
    return this.props.children; // won't work because render can't return an array of elements
    return React.Children.only(this.props.children); // won't work because I may need more than one child rendered
  }
```

I realize I can pass a React component to my loader and tell it to wrap any of it's content into that component, but I think that's quite messy.

**TL;DR** I want a component that can do all the logic of the current component implementation, but without rendering itself (but still rendering children).

**Comment by jimfb** [comment]
What you want is https://github.com/facebook/react/issues/2127

You can modify your function above to be:
```
  render: function() {
    try { return React.Children.only(this.props.children); }
    catch(e) { return <div>{this.props.children}</div> }
  }
```
Which isn't perfect by any means (for one thing, try-catch blocks are a slow path), but it's probably the closest you can come without returning multiple elements.

Closing as a duplicate of https://github.com/facebook/react/issues/2127, feel free to continue the discussion here.

**Comment by mrozbarry** [comment]
Well, this isn't quite a duplicate.  What I want is to not have to render the <div> portion at all.  In effect, the component is just a transparent pass through to do logical processing and to change the render pipeline when data has been loaded or is loading.

**Comment by zpao** [comment]
You're right that it's not a straight duplicate but making that work would allow what you want. And the root of what you're describing is to return 1 or more items from render, so I think it's covered by the same thing.

In our experience doing this we've effectively said these container components must have a single child and that's the only way they will work, so they use `React.children.only`. Relay makes extensive use of this idea. See [react-static-container](https://github.com/reactjs/react-static-container/blob/master/src/StaticContainer.react.js) for example.

That doesn't work with your table example so that's a trickier case and we would probably just say "make the whole table a standalone component / wrapped in the container".
